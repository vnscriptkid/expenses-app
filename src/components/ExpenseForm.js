import React, { Component } from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

export default class ExpenseForm extends Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? props.expense.amount.toString() : '',
            error: '',
            date: props.expense ? moment(props.expense.createdAt) : moment(),
            focused: false
        }
        this.onDescriptionChange = this.onDescriptionChange.bind(this);
        this.onNoteChange = this.onNoteChange.bind(this);
        this.onAmountChange = this.onAmountChange.bind(this);
        this.onFormSubmit = this.onFormSubmit.bind(this);
    }

    onDescriptionChange(e) {
        const description = e.target.value;
        this.setState(() => ({ description }))
    }

    onNoteChange(e) {
        const note = e.target.value;
        this.setState(() => ({ note }));
    }

    onAmountChange(e) {
        const amount = e.target.value;
        this.setState(() => ({ amount }));
    }

    onFormSubmit(e) {
        e.preventDefault();
        if (this.state.description && parseFloat(this.state.amount) > 0) {
            this.props.onFormSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount),
                note: this.state.note,
                createdAt: this.state.date.valueOf()               
            })
            this.setState(() => ({
                error: '',
                description: '',
                amount: '',
                note: ''
            }))
        } else {
            this.setState(() => ({ error: 'You need to provide proper values for description and amount' }))
        }
    }

    render() {
        return (
            <form onSubmit={this.onFormSubmit}>
                <input
                    type="text"
                    value={this.state.description}
                    placeholder="Description"
                    onChange={this.onDescriptionChange}
                />
                <input
                    type="number"
                    placeholder="Amount"
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                />
                <textarea
                    placeholder="Note something"
                    value={this.state.note}
                    onChange={this.onNoteChange}
                >
                </textarea>
                <SingleDatePicker
                    date={this.state.date} // momentPropTypes.momentObj or null
                    onDateChange={date => this.setState({ date })} // PropTypes.func.isRequired
                    focused={this.state.focused} // PropTypes.bool
                    onFocusChange={({ focused }) => this.setState({ focused })} // PropTypes.func.isRequired
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
                <br />
                <button>Add Expense</button>
                {this.state.error && <h3 style={{ color: 'red' }}>{this.state.error}</h3>}
            </form>
        )
    }
}
