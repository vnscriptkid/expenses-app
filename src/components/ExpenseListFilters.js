import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';
import moment from 'moment';
import 'react-dates/initialize';
import 'react-dates/lib/css/_datepicker.css';

class ExpenseListFilters extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            startDate: this.props.filters.startDate,
            endDate: this.props.filters.endDate,
            focusedInput: null
        }
    }

    render() {
        return (
            <div>
                <input type="text" value={this.props.filters.text} onChange={(e) => {
                    this.props.dispatch(setTextFilter(e.target.value));
                }} />

                <select onChange={(e) => {                    
                    if (e.target.value === 'date') {
                        this.props.dispatch(sortByDate());
                    } else if (e.target.value === 'amount') {
                        this.props.dispatch(sortByAmount());
                    }
                }}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                
                <DateRangePicker
                    startDate={this.state.startDate} // momentPropTypes.momentObj or null,
                    endDate={this.state.endDate} // momentPropTypes.momentObj or null,
                    onDatesChange={({ startDate, endDate }) => this.setState({ startDate, endDate })} // PropTypes.func.isRequired,
                    focusedInput={this.state.focusedInput} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                    onFocusChange={focusedInput => this.setState({ focusedInput })} // PropTypes.func.isRequired,
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
        
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}

// export default ExpenseListFilters;
export default connect(mapStateToProps)(ExpenseListFilters);