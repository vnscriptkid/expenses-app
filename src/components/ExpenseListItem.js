import React from 'react';
// import {connect} from 'react-redux';
// import { removeExpense } from '../actions/expenses';
import numeral from 'numeral';
import moment from 'moment';
import { Link } from 'react-router-dom';

export const ExpenseListItem = ({ id, description, amount, createdAt, count }) => {
    return (
        <div>
            <Link to={`/edit/${id}`}>
                <h3 style={{color: 'green'}}>{description}</h3>
            </Link>
            <p>{count} . {description}</p>
            <p> Amount {amount} - Created at {moment(createdAt).format('MMM DD YYYY')}</p>                               
        </div>
    )
}

// export default connect()(ExpenseListItem);
export default ExpenseListItem;
