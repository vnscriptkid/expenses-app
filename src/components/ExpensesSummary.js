import React from 'react';
import getTotalExpenses from '../selectors/expenses-total';
import expensesSelected from '../selectors/expenses';
import { connect } from 'react-redux';

export class ExpensesSummary extends React.Component {
    render() {
        const expenseWord = this.props.countExpenses <= 1 ? 'expense' : 'expenses';
        return (
            <div>
                <p>There are {this.props.countExpenses} {expenseWord} visible </p>
                <p>The total are {this.props.totalExpenses}</p>                
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    const visibleExpenses = expensesSelected(state.expenses, state.filters);
    return {
        countExpenses: visibleExpenses.length,
        totalExpenses: getTotalExpenses(visibleExpenses)
    }
}

export default connect(mapStateToProps)(ExpensesSummary);