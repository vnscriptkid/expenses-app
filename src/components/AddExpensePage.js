import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses';

const AddExpensePage = (props) => (
    <div>
        <h2>ADD EXPENSE PAGE</h2>
        <ExpenseForm onFormSubmit={(expense) => {
            props.dispatch(addExpense(expense));
            props.history.push('/');
            console.log(props.history);
            console.log(expense);
        }}/>
    </div>
)

export default connect()(AddExpensePage);
// export default AddExpensePage;
