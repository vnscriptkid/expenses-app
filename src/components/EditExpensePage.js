import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    constructor(props) {
        super(props);
        this.onEditSubmit = this.onEditSubmit.bind(this);
        this.onRemoveClick = this.onRemoveClick.bind(this);
    }

    onEditSubmit( expense ) {
        this.props.editExpense(
            this.props.expense.id,
            {
                description: expense.description,
                note: expense.note,
                amount: expense.amount,
                createdAt: expense.createdAt
            }
        )
        this.props.history.push('/');
    }

    onRemoveClick() {
        this.props.removeExpense({ id: this.props.expense.id });
        this.props.history.push('/');
    }

    render() {
        return (
            <div>
                <button onClick={this.onRemoveClick}>REMOVE EXPENSE</button>
                <ExpenseForm
                    functionButton="EDIT EXPENSE"
                    expense={this.props.expense}
                    onFormSubmit={this.onEditSubmit}
                />
            </div>
        )
    }
}

// const EditExpensePage = (props) => {
//     return (
//         <div>      
//             <button onClick={() => {
//                 props.dispatch(removeExpense({ id: props.expense.id }));
//                 props.history.push('/');
//             }}>REMOVE EXPENSE</button>     
//             <ExpenseForm 
//                 expense = {props.expense}
//                 onFormSubmit = {(expense) => {
//                     props.dispatch(editExpense(
//                         props.expense.id,
//                         {
//                             description: expense.description,
//                             note: expense.note,
//                             amount: expense.amount,
//                             createdAt: expense.createdAt
//                         }
//                     ))
//                     props.history.push('/');
//                 }}
//             />
//         </div>
//     )
// }


const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    }
}

const mapDispatchToProps = (dispatch) => ({
    editExpense: (id, updates) => dispatch(editExpense(id, updates)),
    removeExpense: ( idObject ) => dispatch(startRemoveExpense( idObject ))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
// export default EditExpensePage;