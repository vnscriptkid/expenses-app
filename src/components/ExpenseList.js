import React from 'react';
import ExpenseListItem from './ExpenseListItem';
import {connect} from 'react-redux';
import ExpensesSelect from '../selectors/expenses';

export const ExpenseList = (props) => {      
    return (
        <div>
            <h3>Expense List</h3>
            { 
                props.expenses.length > 0 ? (
                    <p>There are {props.expenses.length} item(s) in the expense list</p>,
                    props.expenses.map((expense, index) => (                
                        <ExpenseListItem 
                            id={expense.id} 
                            key={expense.id} 
                            count={index+1} 
                            {...expense}
                        />
                    ))
                ) : (
                    <h4>There is no a single one Expense</h4>
                )
            }                                   
        </div>
    );
} 

const mapStateToProps = (state) => {    
    return {
        expenses: ExpensesSelect(state.expenses, state.filters) 
    }
}

// export default ExpenseList;
export default connect(mapStateToProps)(ExpenseList);