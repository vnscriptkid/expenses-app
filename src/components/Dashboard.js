import React from 'react';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseList from './ExpenseList';
import ExpensesSummary from './ExpensesSummary';
import { connect } from 'react-redux';

export const Dashboard = (props) => (
    <div>
        <ExpenseListFilters />
        <ExpensesSummary />
        <ExpenseList />
    </div>
)

export default Dashboard;