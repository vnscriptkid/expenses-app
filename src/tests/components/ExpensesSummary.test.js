import { ExpensesSummary } from '../../components/ExpensesSummary';
import { shallow } from 'enzyme';
import React from 'react';

test('should render ExpensesSummary correctly with 1 expense', () => {
    const wrapper = shallow( <ExpensesSummary countExpenses={1} totalExpenses={100}/> );
    expect(wrapper).toMatchSnapshot();
})

test('should render ExpensesSummary correctly with multiple expenses', () => {
    const wrapper = shallow( <ExpensesSummary countExpenses={100} totalExpenses={200}/> );
    expect(wrapper).toMatchSnapshot();
})

