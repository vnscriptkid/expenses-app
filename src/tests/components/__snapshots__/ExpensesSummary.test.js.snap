// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary correctly with 1 expense 1`] = `
<div>
  <p>
    There are 
    1
     
    expense
     visible 
  </p>
  <p>
    The total are 
    100
  </p>
</div>
`;

exports[`should render ExpensesSummary correctly with multiple expenses 1`] = `
<div>
  <p>
    There are 
    100
     
    expenses
     visible 
  </p>
  <p>
    The total are 
    200
  </p>
</div>
`;
